### PR TITLE
Add on_stall_conditions_changed to EventListener

### PR DESCRIPTION
**Pull request by solotzg**
Add `on_stall_conditions_changed` to `EventListener`, so that it could be shown in Grafana.
`triggered_writes_stop` and `triggered_writes_slowdown` are also exposed as interfaces.

### CODE REVIEW

File: src/event_listener.rs [hunk 1]
```diff
@@ -133,6 +134,27 @@ impl IngestionInfo {
     }
 }
 
+pub struct WriteStallInfo(DBWriteStallInfo);
+pub type WriteStallCondition = DBWriteStallCondition;
```
Comment (overvenus) [inline src/event_listener.rs]: What about `tuple struct`?

File: librocksdb_sys/crocksdb/c.cc [hunk 2]
```diff
@@ -1845,11 +1881,22 @@ struct crocksdb_eventlistener_t : public EventListener {
...
       void*, crocksdb_t*, const crocksdb_externalfileingestioninfo_t*);
+  void (*on_stall_conditions_changed)(void*, const crocksdb_writestallinfo_t*);
 
   virtual void OnFlushCompleted(DB* db, const FlushJobInfo& info) {
     crocksdb_t c_db = {db};
     on_flush_completed(state_, &c_db,
                        reinterpret_cast<const crocksdb_flushjobinfo_t*>(&info));
+
+    {
+      #ifdef TEST_STALL_CONDITIONS_CHANGED
```
Comment (huachaohuang) [inline librocksdb_sys/crocksdb/c.cc]: Seems very hacky, RocksDB provides a `PauseBackgroundWork()` function, maybe you can try to use it and flush some memtables to see if you can trigger a write stall.

File: src/event_listener.rs [hunk 3]
```diff
@@ -133,6 +142,26 @@ impl IngestionInfo {
     }
 }
 
+pub struct WriteStallInfo(DBWriteStallInfo);
+
+#[derive(Copy, Clone)]
+pub struct WriteStallConditionChange {
```
Comment (huachaohuang) [inline src/event_listener.rs]: Where is this used?

Comment (solotzg) [reply src/event_listener.rs]: I forgot to remove it.

**Comment by solotzg** [comment]
@huachaohuang PTAL.

**Comment by huachaohuang** [comment]
@DorianZheng PTAL

File: librocksdb_sys/crocksdb/c.cc [hunk 4]
```diff
@@ -1834,6 +1850,26 @@ crocksdb_externalfileingestioninfo_table_properties(
...
 
+/* External write stall info */
+extern C_ROCKSDB_LIBRARY_API const char*
+crocksdb_writestallinfo_cf_name(
+    const crocksdb_writestallinfo_t* info, size_t* size) {
+  *size = info->rep.cf_name.size();
+  return info->rep.cf_name.data();
+}
+
+const crocksdb_writestallcondition_t* crocksdb_writestallinfo_cur(
```
Comment (DorianZheng) [inline librocksdb_sys/crocksdb/c.cc]: Can we add test case for `cur` and `prev`?

File: tests/test_event_listener.rs [hunk 5]
```diff
@@ -85,6 +85,63 @@ impl EventListener for EventCounter {
...
+        assert!(!info.cf_name().is_empty());
+        self.flush.fetch_add(1, Ordering::SeqCst);
+        self.triggered_writes_slowdown
+            .fetch_add(info.triggered_writes_slowdown() as usize, Ordering::SeqCst);
+        self.triggered_writes_stop
+            .fetch_add(info.triggered_writes_stop() as usize, Ordering::SeqCst);
+    }
+
+    fn on_stall_conditions_changed(&self, info: &WriteStallInfo) {
+        assert!(!info.cf_name().is_empty());
```
Comment (DorianZheng) [inline tests/test_event_listener.rs]: Can we assert something like 
```
assert_eq!(info.cf_name(), expected_name);
```

File: tests/test_event_listener.rs [hunk 6]
```diff
@@ -85,6 +85,63 @@ impl EventListener for EventCounter {
...
+            format!("{:04}", i).as_bytes(),
+        ).unwrap();
+        db.flush(true).unwrap();
+    }
+    let flush_cnt = counter.flush.load(Ordering::SeqCst);
+    assert_ne!(flush_cnt, 0);
+    let stall_conditions_changed_num = counter.stall_conditions_changed_num.load(Ordering::SeqCst);
+    let triggered_writes_slowdown = counter.triggered_writes_slowdown.load(Ordering::SeqCst);
+    let triggered_writes_stop = counter.triggered_writes_stop.load(Ordering::SeqCst);
+    assert_ne!(stall_conditions_changed_num, 0);
```
Comment (DorianZheng) [inline tests/test_event_listener.rs]: Can we assert something like:
```
assert_eq(stall_conditions_changed_num, expected_num);
```
same for two assert below

File: tests/test_event_listener.rs (hunk 5 above)
Comment (solotzg) [reply tests/test_event_listener.rs]: Created with default config, names of those ColumnFamily are all `default`.

File: librocksdb_sys/crocksdb/c.cc (hunk 4 above)
Comment (solotzg) [reply librocksdb_sys/crocksdb/c.cc]: I will add tests.

File: tests/test_event_listener.rs (hunk 5 above)
Comment (solotzg) [reply tests/test_event_listener.rs]: @DorianZheng I have updated the test by creating and using another cf.

**Comment by solotzg** [comment]
@huachaohuang @DorianZheng PTAL.

**Pull request merged**